kubeadm-setup: add page about control plane flags

neolit123 opened the pull request:
This page adds instructions on how to use the kubeadm config
to pass flags to control plane components.

The provided examples are pretty basic. Later this can be expanded with
actual user stories.

Updates https://github.com/kubernetes/kubeadm/issues/849

/cc @fabriziopandini @luxas @timstclair @kubernetes/sig-cluster-lifecycle-pr-reviews  
/assign @Bradamant3 

notes:
- it's against the master branch as discussed on the kubeadm office hours from 13th June.
- i've ended up picking `setup/independent` as the location for this. 
- let me know of changes that i need to make - e.g. hugo formatting?
- this obviously has to be extended with actual user content but we plan to do that in time.

luxas commented on content/en/docs/setup/independent/control-plane-flags.md:
````diff
@@ -0,0 +1,59 @@
+Example usage:
+```yaml
+schedulerExtraArgs:
+  address: 192.168.0.104
````
I'd do 0.0.0.0 as a generic reasonable value for this

luxas commented on content/en/docs/setup/independent/control-plane-flags.md:
```diff
@@ -0,0 +1,59 @@
+schedulerExtraArgs:
+  address: 192.168.0.104
+  config: /home/johndoe/schedconfig
+  master: 192.168.0.103
```
this should be a URL, e.g. `https://my-api-server.domain.com:6443`

luxas replied:
but preferably, I'd set `kubeconfig` to something here instead...

luxas commented on content/en/docs/setup/independent/control-plane-flags.md:
````diff
@@ -0,0 +1,59 @@
+```yaml
+schedulerExtraArgs:
+  address: 192.168.0.104
+  config: /home/johndoe/schedconfig
````
.yaml

luxas commented on content/en/docs/setup/independent/control-plane-flags.md:
````diff
@@ -0,0 +1,59 @@
+```yaml
+controllerManagerExtraArgs:
+  cluster-signing-key-file: /home/johndoe/keys/ca.key
+  bind-address: 192.168.0.101
````
0.0.0.0

luxas commented on content/en/docs/setup/independent/control-plane-flags.md:
```diff
@@ -0,0 +1,59 @@
+controllerManagerExtraArgs:
+  cluster-signing-key-file: /home/johndoe/keys/ca.key
+  bind-address: 192.168.0.101
+  deployment-controller-sync-period: 50
```
this is a duration right? 50s?

timothysc replied:
yes.

luxas commented on content/en/docs/setup/independent/control-plane-flags.md:
````diff
@@ -0,0 +1,59 @@
+```yaml
+apiServerExtraArgs:
+  advertise-address: 192.168.0.103
+  allow-privileged: true
````
is deprecated I think... some other common value like `enable-admission-control`?

timothysc replied:
The most common uses are:

adding admission controllers 
configuring audit log (soon to be webhook)

luxas commented on content/en/docs/setup/independent/control-plane-flags.md:
```diff
@@ -0,0 +1,59 @@
+- `ControllerManagerExtraArgs`
+- `SchedulerExtraArgs`
+
+The fields can contain a list of `key: value` pairs. To override flags for a specific control plane component just include one of the fields above and start adding flags for it.
```
you can add a `(it's serialized as a map[string]string`, like Kubernetes labels or annotations)`

timothysc replied:
s/just/

Bradamant3 replied:
What value for a user does the content that luxas suggests add? This seems like an implementation detail that does not belong here. We've already got a larger confusion in the kubeadm docs between what a user needs and what a dev needs; let's not perpetuate it here.

Bradamant3 replied:
Larger scale edit:

These fields consist of `key: value` pairs. To override a flag for a control plane component: 

1. Add the appropriate field to your configuration.
1. Add the flags to override to the field.

luxas commented on content/en/docs/setup/independent/control-plane-flags.md:
````diff
@@ -0,0 +1,59 @@
+
+Example usage:
+```yaml
+apiServerExtraArgs:
````
include full typemeta information in every example please

luxas commented on content/en/docs/setup/independent/control-plane-flags.md:
```diff
@@ -0,0 +1,59 @@
+- fabriziopandini
+- luxas
+- timothysc 
+title: Customizing control plane flags with kubeadm
```
`Customizing control plane configuration with kubeadm` is more generic and future-proof

luxas commented:
/assign

timothysc commented:
Is this the right location?

neolit123 commented:
@luxas, @timstclair thanks for the reviews. i will update this tomorrow.

> Is this the right location?

it felt right to me to under setup and not tasks or cmd reference.
i'm not sure; maybe collect more opinions?

Bradamant3 commented:
For location, see https://docs.google.com/document/d/1-pAxBmV7c0DfbJVU33oK4Cozd99Z_16HX1-MLZ-A6OY/edit?usp=sharing

Bradamant3 commented on content/en/docs/setup/independent/control-plane-flags.md:
```diff
@@ -0,0 +1,59 @@
+
+{{< note >}}
+In later versions of the kubeadm configuration these fields will be replaced by usage of a ComponentConfig and a ConfigMap.
+{{< /note >}}
```
Remove entire note. Future statements are dangerous, and of no use to current users. (Again, this is one of those user/developer confusions.)

Bradamant3 commented on content/en/docs/setup/independent/control-plane-flags.md:
```diff
@@ -0,0 +1,59 @@
+
+## APIServer flags
+
+The list of possible APIServer flags and their descriptions can be found [here](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-apiserver/).
```
edit to:
"For details, see [the reference documentation for kube-apiserver](/docs/reference/command-line-tools-reference/kube-apiserver/)."

(note formation of link also)

Bradamant3 commented on content/en/docs/setup/independent/control-plane-flags.md:
```diff
@@ -0,0 +1,59 @@
+
+## ControllerManager flags
+
+The list of possible ControllerManager flags and their descriptions can be found [here](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-controller-manager/).
```
See comment at line 27 -- follow the same pattern here.

Bradamant3 commented on content/en/docs/setup/independent/control-plane-flags.md:
```diff
@@ -0,0 +1,59 @@
+
+## Scheduler flags
+
+The list of possible Scheduler flags and their descriptions can be found [here](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-scheduler/).
```
See line 27, follow pattern.

neolit123 commented:
> i will update this tomorrow.

i'm going to have to delay the edits for the weekend or monday.

neolit123 commented:
hi, @timstclair @luxas @Bradamant3 
i've added the changes. LMK if this better and i will rebase / squash.

in terms of location i'm leaving this as is, unless someone objects?

@Bradamant3 thanks for taking the time to write the document:
https://docs.google.com/document/d/1-pAxBmV7c0DfbJVU33oK4Cozd99Z_16HX1-MLZ-A6OY/edit?usp=sharing

do we have a tracking issue for the kubeadm docs folder restructuring and if not do you want me to create one? also in which repo `k/website` or `k/kubeadm`?

timothysc commented:
@neolit123 - only comment is the addition of the meta-data @luxas mentioned earlier, other

neolit123 commented:
@timothysc 
>only comment is the addition of the meta-data @luxas mentioned earlier, other

i've included `apiVersion` and `kind` from the [apimachinery/pkg/apis/meta/v1/TypeMeta](https://github.com/kubernetes/apimachinery/blob/master/pkg/apis/meta/v1/types.go#L41) embedded struct. is something else needed?

timothysc commented:
@neolit123 I think adding: 

``` 
metadata: 
   name: 1.11-sample
``` 
 is useful to the readers w/possible link to https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.10/#objectmeta-v1-meta optional params.

neolit123 commented:
added the `metadata` (`metav1.ObjectMeta`) field examples as @timothysc suggested.
added link to godocs, which should include `metav1.ObjectMeta` and the latest config when the new generated docs are up.

neolit123 commented:
squashed commits. waiting on approval.

Bradamant3 commented:
approved review changes, not adding hold but in case anybody else is looking here: author is still fixing a typo and some template problems. I don't have commit access (GH ftw ...)

Bradamant3 commented:
/lgtm
/approve
